content/courses/ada-in-practice/chapters/raii.rst

```diff
@@ -0,0 +1,480 @@
+.. _Ada_In_Practice_RAII:
+
+Resource Acquisition Is Initialization (RAII)
+=============================================
+
+.. include:: ../../global.txt
+
+Motivation
+----------
+
+In order for the expected semantics to be obtained, some types require clients
+to follow a specific protocol when calling the type's operations. Furthermore,
+failing to follow the protocol can cause system-wide ill effects.
+
+For example, concurrency abstractions such as mutexes provide the mutually
+exclusive access necessary to prevent the race conditions that arise when
+competing concurrent threads access shared resources. These mutex objects must
+be 1) both acquired and released, 2) by every thread accessing that shared
+resource, 3) at the right places in the source code, and 4) in the proper
+order. Failure to acquire the mutex prior to accessing the shared resource
+leads to race conditions, and failure to release it can lead to deadlocks.
+Ensuring the mutex is released is complicated by the possibility of exceptions
+raised after the lock is acquired.
+
+Although concurrency is a prime example, the issue is general in nature. We
+will continue with the concurrency context for the sake of discussion.
+
+Like the classic *monitor* concept (:footcite:p:`1977:hansen`,
+:footcite:p:`1973:hansen`, :footcite:p:`1974:hoare`) on which they are based.
+(In the works cited, Hoare's contribution was equally important, but Hansen's
+contributions were reified in
+:wikipedia:`Concurrent Pascal <Concurrent_Pascal>`, a concrete programming
+language.), Ada defines a protected object (PO) as a concurrency construct that
+is higher-level and more robust than mutexes and semaphores. Those advantages
+accrue because the bodies of the protected operations implement only the
+functional requirements, with guaranteed mutually exclusive access to the
+protected data. The underlying implementation, i.e., the run-time library,
+provides the necessary exclusive access, and also thread management. As a
+result, the lock is guaranteed to be acquired, and likewise, guaranteed to be
+relinquished, even in the face of exceptions.
+
```

Comment:
Let's change the sentences that end the paragraph above (starting at line 35) to the following:
accrue because the bodies of the protected operations are only responsible for implementing the
functional requirements. The underlying run-time library is responsible for 
implementing the mutually exclusive access, and also thread management. As a
result, the source code is much simpler and is robust even in the face of exceptions.